Experiment: Render timing pipeline in SVG

eth3lbert opened the pull request:
### What does this PR try to resolve?

This PR adds an SVG renderer and allows users to choose between Canvas and SVG for rendering charts, which should hopefully resolve #8850.
<img width="1405" height="937" alt="image" src="https://github.com/user-attachments/assets/760a1a61-72b3-45bc-b36a-bb581e50c0a6" />



### Additional information

This does not guarantee a pixel-perfect match from the previous canvas-based solution, but I maintain it stays as close as possible.

ehuss commented:
For context, the original version of this used SVG. However, it was switched to canvas due to severe rendering performance problems on some systems and browsers. Has this been thoroughly tested on multiple browsers, and on systems with limited video cards, and different platforms?

eth3lbert commented:
> For context, the original version of this used SVG.

Yeah, I noticed this part.

> Has this been thoroughly tested on multiple browsers, and on systems with limited video cards, and different platforms?

No, I haven't tested it with legacy devices or systems, only the newest Firefox and Chrome. Testers and feedback are welcome :D

weihanglo commented:
> However, it was switched to canvas due to severe rendering performance problems on some systems and browsers

Any chance to recall those legacy systems and browsers?

ehuss commented:
> Any chance to recall those legacy systems and browsers?

It was probably a 2018-era Intel integrated UHD Graphics 630 or AMD Radeon Pro 560X with an Intel Core i7 (macbook). I think Firefox was one of the worst in performance, but Safari and Chrome were also pretty bad. I believe Firefox has improved its SVG rendering since then.

eth3lbert commented:
> I believe Firefox has improved its SVG rendering since then.

For Firefox, although the SVG renders without issues, I encountered significant frame drops while scrolling on an M1 device when experimenting with this SVG solution for `crates.io` (which has `619 units` in total). This is primarily due to those dependency lines. If I hide those dependency lines, everything works fine.

I'm not sure if this performance issue is solvable without modifying the presentation. If we only rendered dependency lines for on-screen blocks, the performance would improve significantly, although I'm not sure if this approach is desirable.

eth3lbert commented:
> That said, it might be a bit easier to implement now since the `RenderContext` provide an separation of data and presentation layer: #16282. We could probably have an unstable SVG timing report format for people to test :)

Thanks for the information, I'll take a look in the near future :D

> Or we could have a button for people to switch between two modes 😆

Is this referring to canvas and SVG rendering, or to panning and zooming?

eth3lbert commented:
> For switching between SVG/canvas

I've added this now! The CI failed, but it shouldn't be related to this PR.  The scrolling jank in Firefox is also fixed on my end, so please give it a try and let me know what you think. :D

eth3lbert commented:
I also found that highlighting connected units on hover is less useful when there are many units. They might not all fit on the screen, which means you have to scroll a long way to find the connections. Plus, you might accidentally hover over other units while scrolling. Because of this, a different interface would be more helpful for showing these relations. For example, clicking a unit to show its connections in a side panel might provide a better UX.

eth3lbert commented on src/cargo/core/compiler/timings/report.rs:
```diff
@@ -803,18 +808,31 @@ static HTML_CANVAS: &str = r#"
+            <input type="radio" name="renderer" value="svg" />
+            SVG
+        </label>
+    </td>
```
I plan to make the SVG renderer the default, since it outperformed Canvas in my testing across Firefox, Chrome, and Safari!

weihanglo commented on src/cargo/core/compiler/timings/report.rs:
> I also found that highlighting connected units on hover is less useful when there are many units. They might not all fit on the screen, which means you have to scroll a long way to find the connections. Plus, you might accidentally hover over other units while scrolling. Because of this, a different interface would be more helpful for showing these relations. For example, clicking a unit to show its connections in a side panel might provide a better UX.

Agree with side panel solution. That sounds pretty approachable.
I am not sure if it is easy to implement though, as currently we don't use any Web Frontend frame work, which may make it a bit hard to maintain

weihanglo commented on src/cargo/core/compiler/timings/report.rs:
I just tried with `CARGO_BUILD_ANALYSIS_ENABLED=true rustup run nightly target/debug/cargo build -Zsection-timings -Zbuild-analysis && target/debug/cargo report timings -Zbuild-analysis --open`. The performance enhancement was so huge!

* Google Chrome: Version 143.0.7499.170 (Official Build) (arm64)
* Safari: Version 26.2 (20623.1.14.18.4)
* Firefox: 145.0 (aarch64)

1. Tried set the scale factor to 9.
2. Chose the canvas renderer
    * Observed on Chrome and Safari, it was laggy and stopped rendering at some point.
    * Observed on Firefox it worked.
3. Switched to the SVG renderer
    * All of them works well.
    * On Firefox the dependency line drawing was faster than the canvas renderer.


Here is the timing HTML: [cargo-timing.html](https://github.com/user-attachments/files/24339624/cargo-timing.html)

weihanglo reviewed (approved):
Thanks for working on this. The PR is so nice for both refactors and the SVG implementation. 

I assume this doesn't need any team FCP:

* It offers a switch between canvas and SVG rendering options. If anything broken, people can always switch back.
* The default is still canvas.
* The HTML report UI is for human not for machine, so we have more flexibility to change it.


I am excited to merge this right now! Really appreciate it!

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/cargo/15091/aaa648864cf99ff3668bea69e6d1e2824886abda..d4e9a7abf5b1e1e44213c8c49615d2daa84cbfe4)*
